File: frontend/src/api/titles.ts
import { apiFetch } from "./client";

export type Genre = { id: string; name: string };

export type Credit = {
  name: string;
  credit_type: string;
  job: string | null;
  character: string | null;
  billing_order: number | null;
  profile_url: string | null;
};

export type Title = {
  id: string;
  media_type: string;
  name: string;
  overview: string | null;
  release_date: string | null;
  runtime: number | null;
  popularity: number;
  vote_average: number;
  poster_path: string | null;
  backdrop_path: string | null;
  original_language: string | null;
  genres: Genre[];
  poster_url: string | null;
};

export type TitleDetail = Title & {
  credits: Credit[];
  keywords: string[];
};

export type ProviderOffer = {
  provider_id: number;
  name: string;
  logo_url: string | null;
  display_priority: number;
};

export type WhereToWatch = {
  region: string;
  link: string | null;
  flatrate: ProviderOffer[];
  free: ProviderOffer[];
  ads: ProviderOffer[];
  rent: ProviderOffer[];
  buy: ProviderOffer[];
  available: boolean;
  attribution: string;
  source: string;
};

export type Reason = {
  /** Stable reason kind (because_you_liked, taste_blend, same_director, …). */
  code: string;
  /** User-facing explanation sentence. */
  message: string;
  /** Optional structured support (liked_titles, directors, tones, …). */
  evidence: Record<string, unknown>;
};

export type RecommendationItem = {
  title: Title;
  score: number;
  reasons: Reason[];
};

/** Onboarding / interaction actions that move (or don't move) taste. */
export type OnboardingAction =
  | "haven't_seen"
  | "not_interested"
  | "rate_1"
  | "rate_2"
  | "rate_3"
  | "rate_4";

export type OnboardingReaction = {
  title_id: string;
  action: OnboardingAction;
};

export function getForYou(accessToken: string, limit = 20) {
  return apiFetch<{ items: RecommendationItem[] }>(
    `/recommendations/for-you?limit=${limit}`,
    {},
    accessToken,
  );
}

export function getTitle(accessToken: string, titleId: string) {
  return apiFetch<TitleDetail>(`/titles/${titleId}`, {}, accessToken);
}

export function getSimilarTitles(accessToken: string, titleId: string, limit = 12) {
  return apiFetch<Title[]>(
    `/titles/${titleId}/similar?limit=${limit}`,
    {},
    accessToken,
  );
}

export function getWhereToWatch(
  accessToken: string,
  titleId: string,
  region?: string,
) {
  const qs = region ? `?region=${encodeURIComponent(region)}` : "";
  return apiFetch<WhereToWatch>(
    `/titles/${titleId}/where-to-watch${qs}`,
    {},
    accessToken,
  );
}

export function getOnboardingCards(
  accessToken: string,
  opts?: { limit?: number; exclude?: string[] },
) {
  const params = new URLSearchParams();
  if (opts?.limit) params.set("limit", String(opts.limit));
  for (const id of opts?.exclude ?? []) {
    params.append("exclude", id);
  }
  const qs = params.toString();
  return apiFetch<{ items: Title[] }>(
    `/onboarding/cards${qs ? `?${qs}` : ""}`,
    {},
    accessToken,
  );
}

export function completeOnboarding(
  accessToken: string,
  reactions: OnboardingReaction[],
) {
  return apiFetch<import("./auth").User>(
    "/onboarding/complete",
    {
      method: "POST",
      body: JSON.stringify({ reactions }),
    },
    accessToken,
  );
}

export type InteractionEventType =
  | "like"
  | "dislike"
  | "watchlist"
  | "not_interested"
  | "clear"
  | "skip"
  | "view"
  | "haven't_seen"
  | "rate_1"
  | "rate_2"
  | "rate_3"
  | "rate_4"
  | "watched"
  | "watched_liked"
  | "watched_disliked";

export function interact(
  accessToken: string,
  titleId: string,
  event_type: InteractionEventType,
) {
  return apiFetch<void>(
    `/titles/${titleId}/interactions`,
    {
      method: "POST",
      body: JSON.stringify({ event_type }),
    },
    accessToken,
  );
}

export function searchTitles(accessToken: string, q: string, limit = 24) {
  return apiFetch<Title[]>(
    `/titles/search?q=${encodeURIComponent(q)}&limit=${limit}`,
    {},
    accessToken,
  );
}

export function getWatchlist(accessToken: string) {
  return apiFetch<Title[]>("/watchlist", {}, accessToken);
}

export function getCatalogStatus(accessToken: string) {
  return apiFetch<{
    title_count: number;
    with_embeddings: number;
    ready_for_onboarding: boolean;
  }>("/catalog/status", {}, accessToken);
}

File: backend/app/api/schemas/titles.py
from __future__ import annotations

from datetime import date
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class GenreOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str


class ReasonOut(BaseModel):
    """Human-readable explanation for a recommendation.

    ``message`` is the user-facing line. ``evidence`` holds structured
    detail for UI polish / debugging (liked title names, directors, tones…).
    """

    code: str = Field(
        description=(
            "Reason kind: because_you_liked | taste_blend | similar_style | "
            "same_director | similar_cast | similar_themes | similar_origin | "
            "similar_language | similar_era | shared_genre | taste_similarity | "
            "genre_fit | discovery"
        )
    )
    message: str = Field(description="Thoughtful, specific sentence shown to the user")
    evidence: dict = Field(
        default_factory=dict,
        description="Structured support: liked_titles, directors, tones, keywords, genres, …",
    )


class TitleSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    media_type: str
    name: str
    overview: str | None
    release_date: date | None
    runtime: int | None
    popularity: float
    vote_average: float
    poster_path: str | None
    backdrop_path: str | None
    original_language: str | None
    genres: list[GenreOut] = Field(default_factory=list)

    @property
    def poster_url(self) -> str | None:
        if not self.poster_path:
            return None
        if self.poster_path.startswith("http"):
            return self.poster_path
        return f"https://image.tmdb.org/t/p/w500{self.poster_path}"


class TitleSummaryOut(TitleSummary):
    poster_url: str | None = None

    @classmethod
    def from_title(cls, title) -> TitleSummaryOut:
        data = TitleSummary.model_validate(title)
        poster = None
        if title.poster_path:
            poster = (
                title.poster_path
                if title.poster_path.startswith("http")
                else f"https://image.tmdb.org/t/p/w500{title.poster_path}"
            )
        return cls(**data.model_dump(), poster_url=poster)


class RecommendationItemOut(BaseModel):
    title: TitleSummaryOut
    score: float
    reasons: list[ReasonOut]


class RecommendationSlateOut(BaseModel):
    items: list[RecommendationItemOut]


class InteractionRequest(BaseModel):
    """Record a title interaction.

    Weights and taste effects: ``docs/TASTE_SIGNALS.md`` and
    ``app.domain.taste_signals``.
    """

    event_type: str = Field(
        pattern=(
            "^(like|dislike|watchlist|not_interested|skip|view|"
            "haven't_seen|rate_1|rate_2|rate_3|rate_4)$"
        ),
        description=(
            "Active signals only. "
            "haven't_seen=0 taste; not_interested=mild−; "
            "rate_1…rate_4=Bad…Favorite; watchlist=mild+; "
            "like/dislike=feed shortcuts. See docs/TASTE_SIGNALS.md."
        ),
    )


class OnboardingReaction(BaseModel):
    """One card decision during onboarding.

    Policy (docs/TASTE_SIGNALS.md):
    - haven't_seen — zero taste signal (does not count as a rating)
    - not_interested — mild negative
    - rate_1 … rate_4 — Bad / It's ok / Good / Favorite
    - like / dislike — legacy aliases (mapped to rate_3 / rate_1)
    """

    title_id: UUID
    action: str = Field(
        pattern=(
            "^(haven't_seen|not_interested|rate_1|rate_2|rate_3|rate_4|like|dislike)$"
        )
    )


class OnboardingCompleteRequest(BaseModel):
    reactions: list[OnboardingReaction] = Field(min_length=1, max_length=80)


class OnboardingCardsOut(BaseModel):
    items: list[TitleSummaryOut]


class CatalogStatusOut(BaseModel):
    title_count: int
    with_embeddings: int
    ready_for_onboarding: bool
